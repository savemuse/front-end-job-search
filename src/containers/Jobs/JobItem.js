import React, { Component } from 'react';
import { Card, CardHeader, CardText } from 'material-ui/Card';
import Tags from '../../components/Tags/Tags';

class JobItem extends Component {
  constructor(props) {
    super(props);
    this.state = {
      expanded: false,
    };
  }

  handleExpandChange = (expanded) => {
    this.setState({expanded: expanded});
  };

  render() {
    const { 
      title,
      user,
      created_at,
      updated_at,
      labels, body 
    } = this.props.job;
    const updateDate = updated_at ? new Date(updated_at) : new Date(created_at);
    const updateStr = `${updateDate.toLocaleDateString()} ${updateDate.toLocaleTimeString()}`; 

    return (
      <div className="card-container-item">
        <Card
          expanded={this.state.expanded}
          onExpandChange={this.handleExpandChange}
        >
          <CardHeader
            title={title}
            subtitle={updateStr}
            avatar={user.avatar_url}
            actAsExpander={true}
            showExpandableButton={true}
          />
          <CardText>
            <Tags labels={labels} />
          </CardText>
          <CardText expandable={true}>
            {body}
          </CardText>
        </Card>
      </div>
    );
  }
}

export default JobItem;