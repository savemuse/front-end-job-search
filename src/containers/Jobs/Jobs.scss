.input-container {
  width: 70%;
  margin: 40px auto;
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
}

.card-container {
  width: 80%;
  margin: 40px auto;
  display: flex;
  flex-direction: column;

  &-item {
    width: 100%;
    margin: 30px 0;
  }
}

.text-align-center {
  text-align: center;
  color: #BDBDBD;
}

.empty-container {
  height: 300px;
}