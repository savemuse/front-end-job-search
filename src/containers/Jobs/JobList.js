import React, { Component } from 'react'
import PropTypes from 'prop-types';
import JobItem from './JobItem';

class JobList extends Component {
  componentDidMount() {
    this.props.getJobsRequest();
  }

  render() {
    const { filter_jobs, loading } = this.props;

    if (loading) {
      return (
        <div className="text-align-center empty-container">
          <h3>資料加載中。</h3>
        </div>
      )
    }

    if (filter_jobs.length === 0) {
      return (
        <div className="text-align-center empty-container">
          <h3>您查詢的資料尚無結果，請修改您的查詢條件。</h3>
        </div>
      )
    }

    return (
      <div className="card-container">
        <div className="text-align-center">
          <h3>{`查詢結果共${filter_jobs.length || 0}筆`}</h3>
        </div>
        <ul>
          { filter_jobs.map((job) => (
              <li key={job.id}>
                <JobItem job={job}/>
              </li>
            ))
          }
        </ul>
      </div>
    )
  }
}

JobList.propTypes = {
  loading: PropTypes.bool.isRequired,
  filter_jobs: PropTypes.array.isRequired,
  getJobsRequest: PropTypes.func.isRequired,
}

export default JobList;
