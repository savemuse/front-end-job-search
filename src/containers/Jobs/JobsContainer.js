import { connect } from 'react-redux';
import Jobs from './Jobs';
import { 
  getLabelsRequest,
  filterJobsRequest,
} from '../../redux/modules/jobs/action';

const mapStateToProps = state => {
  const { labels } = state.jobs;
  return {
    labels
  };
};

const mapDispatchToProps = dispatch => {
  return {
    getLabelsRequest: () => {
      dispatch(getLabelsRequest());
    },
    filterJobsRequest: (params) => {
      dispatch(filterJobsRequest(params));
    }
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Jobs);