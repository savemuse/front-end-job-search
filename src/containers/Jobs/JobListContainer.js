import { connect } from 'react-redux';
import JobList from './JobList';
import { getJobsRequest } from '../../redux/modules/jobs/action';

const mapStateToProps = state => {
  const { loading, filter_jobs } = state.jobs;
  return {
    loading,
    filter_jobs,
  };
};

const mapDispatchToProps = dispatch => {
  return {
    getJobsRequest: () => {
      dispatch(getJobsRequest());
    },
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(JobList);