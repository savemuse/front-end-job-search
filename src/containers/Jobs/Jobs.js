import React, { Component } from 'react';
import PropTypes from 'prop-types';
import RaisedButton from 'material-ui/RaisedButton';
import JobListContainer from '../Jobs/JobListContainer';
import Dropdown from '../../components/Dropdown/Dropdown';
import SearchBar from '../../components/SearchBar/SearchBar';
import './Jobs.scss';
const styles = {
  button: {
    margin: '28px 10px 0 10px',
    boxShadow: 'none',
  },
};

class Jobs extends Component {
  constructor(props) {
		super(props);
    this.handleFilterChange = this.handleFilterChange.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleSearch = this.handleSearch.bind(this);
    this.state = {
      text: '',
      filter: null
    };
	}

  componentDidMount() {
    this.props.getLabelsRequest();
  }
  
  handleFilterChange = (value) => {
    this.setState({
      filter: value
    });
  }

  handleChange = (value) => {
    this.setState({
      text: value
    });
  }

  handleSearch = () => {
    const params = {...this.state};
    this.props.filterJobsRequest(params);
  }
  
  render() {
    const { filter } = this.state;
    const { labels } = this.props;

    return (
      <div>
        <div className="input-container">
          <Dropdown
            filter={filter}
            menuList={labels}
            handleFilterChange={this.handleFilterChange}
          />
          <SearchBar
            handleChange={this.handleChange}
          />
          <RaisedButton
            style={styles.button}
            label="SEARCH"
            primary={true}
            onClick={this.handleSearch}
          />
        </div>
        <JobListContainer/>
      </div>
    );
  }
}

Jobs.propTypes = {
  labels: PropTypes.array.isRequired,
  getLabelsRequest: PropTypes.func.isRequired,
  filterJobsRequest: PropTypes.func.isRequired,
}

export default Jobs;