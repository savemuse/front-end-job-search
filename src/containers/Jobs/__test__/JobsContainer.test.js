/* eslint-disable */
jest.mock('../Jobs');

import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import storeFake from '../../../utils/testUtil';
import Jobs from '../Jobs';
import JobsContainer from '../JobsContainer';

describe('container <JobList />', () => {
  let subject;
  let component;
  let jobsComponent;

  beforeEach(() => {
    const store = storeFake({
      jobs: {
        labels: [
          'GIT',
          'unit test',
          'Full Stack',
          '[F] React',
          '[F] Vue.js',
          '[F] D3.js',
          '[B] NodeJS',
          '[B] GraphQL',
          '[B] ROR',
          '年薪百萬',
        ]
      }
    });
    subject = mount(
      <Provider store={store}>
        <JobsContainer />
      </Provider>
    );

    component = subject.find(JobsContainer);
    jobsComponent = component.find(Jobs);
  });

  it('should be called exactly once', () => {
    expect(Jobs.mock.calls.length).toBe(1);
  });

  it('should execute the dispatch function after calling the getLabelsRequest function', () => {
    jobsComponent.props().getLabelsRequest();
    expect(subject.instance().store.dispatch).toBeCalled();
  });

  it('should be same as snapshot', () => {
    expect(toJson(jobsComponent)).toMatchSnapshot();
  });
});
