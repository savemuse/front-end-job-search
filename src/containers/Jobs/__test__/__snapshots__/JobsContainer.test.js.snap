// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`container <JobList /> should be same as snapshot 1`] = `
<Jobs
  filterJobsRequest={[Function]}
  getLabelsRequest={[Function]}
  labels={
    Array [
      "GIT",
      "unit test",
      "Full Stack",
      "[F] React",
      "[F] Vue.js",
      "[F] D3.js",
      "[B] NodeJS",
      "[B] GraphQL",
      "[B] ROR",
      "年薪百萬",
    ]
  }
/>
`;
